### PR TITLE
CMSSW 92X - Adding the possibility to select the PVs before the BeamSpot fit

### PR DESCRIPTION
**Pull request by francescobrivio**
This is a "reprise" of the PR #18201 (now reverted and closed).
Basically we add two parameters in PVfitter in order to be able to apply further selection on PrimaryVertices. All python cfg files have been modified to account for these new parameters (also in DQM/BeamMonitor and in Calibration/TkAlCaRecoProducers).
The default values of the parameters are set so that nothing changes with respect to the previous situation.
We added a further switch to prevent the program from crashing when the needed track collection is not found: in that case the PVselection is not applied.
I also did some "cleaning" (removed "tabs" and useless commented lines) in order to have more or less the same indendation in all the files.

### CODE REVIEW

**Comment by arunhep** [comment]
please test

**Comment by rmanzoni** [comment]
hi,

this PR has 91X in its title and the branch to merge has 91X in its name as well, however it is definitely meant to be in place for the data taking, that is it should enter 92X.

@francescobrivio can you please change the title to 92X?

@franzoni @arunhep might be that this slipped under your radars because of the confusing name, but this is *the* PR for 2017 BS. I see Arun triggered the tests, they went well, please have a look and give +1 (or -1 of course, if that's the case) to move forwards

**Comment by francescobrivio** [comment]
@rmanzoni Title changed to 92X

File: RecoVertex/BeamSpotProducer/src/PVFitter.cc [hunk 1]
```diff
@@ -103,51 +93,44 @@ PVFitter::~PVFitter() {
...
+          {
+              double sumPt=0;
+              for(auto iTrack = pv->tracks_begin(); iTrack != pv->tracks_end(); ++iTrack)
+              {
+                const auto pt = (*iTrack)->pt();
+                sumPt += pt;
+              }
+              if (sumPt < minSumPt_) continue;
+          }
+          catch (...)
```
Comment (ghellwig) [inline RecoVertex/BeamSpotProducer/src/PVFitter.cc]: @francescobrivio this is forbidden by the CMS code rules
Please catch only the exception you want to handle here.

File: RecoVertex/BeamSpotProducer/src/PVFitter.cc [hunk 2]
```diff
@@ -401,12 +373,12 @@ bool PVFitter::runFitter() {
...
       upar.Add("z"     , estZ       , errZ	     , -30.	    , 30.	    ); // 2
-      upar.Add("ex"    , 0.015	    , 0.01  	     , 0.   	    , 10. 	    ); // 3
-      upar.Add("corrxy", 0.   	    , 0.02  	     , -1.  	    , 1.  	    ); // 4
-      upar.Add("ey"    , 0.015	    , 0.01  	     , 0.   	    , 10. 	    ); // 5
-      upar.Add("dxdz"  , 0.   	    , 0.0002	     , -0.1 	    , 0.1 	    ); // 6
-      upar.Add("dydz"  , 0.   	    , 0.0002	     , -0.1 	    , 0.1 	    ); // 7
-      upar.Add("ez"    , 1.   	    , 0.1   	     , 0.   	    , 30. 	    ); // 8
+      upar.Add("ex"    , 0.010	    , 0.0001  	 , 0.   	, 10. 	    ); // 3
+      upar.Add("corrxy", 0.   	    , 0.02  	 , -1.  	, 1.  	    ); // 4
+      upar.Add("ey"    , 0.010	    , 0.0001  	 , 0.   	, 10. 	    ); // 5
```
Comment (ghellwig) [inline RecoVertex/BeamSpotProducer/src/PVFitter.cc]: The PR description says that nothing changes wrt. the previous situation, but here you are changing hard-coded parameter values.
Please clarify the impact of those.

**Review by ghellwig — CHANGES_REQUESTED** [review]
@francescobrivio can you point me to the lines of code or configuration that fix the issue mentioned in https://github.com/cms-sw/cmssw/pull/18201#issuecomment-295599218.
Unfortunately the link from David is now dead as the logs are cleaned up in the meantime.

**Comment by perrotta** [comment]
@francescobrivio 
- Please remove commented out lines from PVFitter.h, as you already did in previous #18201
- Please check otherwise if the track collection exists, as in the first comment of https://github.com/cms-sw/cmssw/pull/18755#pullrequestreview-39418364
- Please either explain or revert the changes in the hard coded parameters, as in the second comment of https://github.com/cms-sw/cmssw/pull/18755#pullrequestreview-39418364

**Comment by franzoni** [comment]
hello @francescobrivio 
perhaps you may answer @ghellwig 's queries ?
We're preparing the PR's for our weekly ORP report and it would be good to proceed.

**Comment by francescobrivio** [comment]
hi @franzoni sorry for the delay. I'm working on it right now.

**Comment by francescobrivio** [comment]
hi @perrotta and @ghellwig 
I removed the commented lines in PVFitter.h and reverted the hard coded parameters in PVFitter.cc,
I just have to fix the try/catch part and then I will push the fixes.
Do you have any suggestion on how to catch a "ProductNotFound" exception in cmssw?

**Comment by makortel** [comment]
@francescobrivio Check beforehand if `edm::Handle` is valid, or `Ref` is not null and is available. Here is an example how we nowadays check in the vertex validation code that the track collection referenced by the vertex collection is available in the `edm::Event` (and ignore the vertex collection if the track collection is missing)
https://github.com/cms-sw/cmssw/blob/master/Validation/RecoVertex/src/PrimaryVertexAnalyzer4PUSlimmed.cc#L1373-L1391

**Comment by francescobrivio** [comment]
Hi @makortel thanks a lot for the tip, it was very useful!

**Comment by ghellwig** [comment]
please test

**Comment by ghellwig** [comment]
@francescobrivio thanks for the fixes, but I believe you still owe me an answer to this:

> @francescobrivio can you point me to the lines of code or configuration that fix the issue mentioned in https://github.com/cms-sw/cmssw/pull/18201#issuecomment-295599218.
Unfortunately the link from David is now dead as the logs are cleaned up in the meantime.

**Comment by francescobrivio** [comment]
sorry I forgot to answer you!
the fix is indeed the try/catch which is now a more correct check on the edm::handle.
I hope this answers your question, if not, please let me know!
________________________________________
Da: Gregor Mittag [notifications@github.com]
Inviato: martedì 23 maggio 2017 18.58
A: cms-sw/cmssw
Cc: Francesco Brivio; Mention
Oggetto: Re: [cms-sw/cmssw] CMSSW 92X PVselections (#18755)

@francescobrivio<https://github.com/francescobrivio> thanks for the fixes, but I believe you still owe me an answer to this:

@francescobrivio<https://github.com/francescobrivio> can you point me to the lines of code or configuration that fix the issue mentioned in #18201 (comment)<https://github.com/cms-sw/cmssw/pull/18201#issuecomment-295599218>.
Unfortunately the link from David is now dead as the logs are cleaned up in the meantime.

**Comment by ghellwig** [comment]
+1

**Comment by ghellwig** [comment]
thanks for the explanation

**Comment by perrotta** [comment]
+1
As in #18201 (already merged and the reverted), this PR basically adds two extra parameters to be used for the fit of the BeamSpot: they can be used for tests, but they are kept by now at default values that fully reproduce the previous behavior of the algo.
With respect to the previous implementation, here a protection is added with a check on the presence of the track collection linked to the PV. Explicitely checked that the previous failing wf 1004 is now ending up correctly.

**Comment by perrotta** [comment]
@francescobrivio : please change the title of this PR to something more descriptive. The previous "Adding the possibility to select the PVs before the BeamSpot fit" would be fine, for example.

**Comment by francescobrivio** [comment]
@perrotta title changed

**Comment by dmitrijus** [comment]
+1

**Pull request merged**